[#45414] [AppSignal] SearchController#index response time is very slow

dombesz opened the pull request:
https://community.openproject.org/work_packages/45414

dombesz committed:
[#45414] [AppSignal] SearchController#index response time is very slow (more than 2 min)

https://community.openproject.org/work_packages/45414

dombesz commented:
Upon analysing the original query we concluded that some of the text fields could be improved by adding [GIN indexes](https://pganalyze.com/blog/gin-index). Later I found out the GIN indexes are not being used in the query planner. The reason is multiple left joins with complex OR conditions can trip the query planner and not use the appropriate indexes ([see this](https://dba.stackexchange.com/questions/204349/postgres-not-using-indexes-when-oring-across-join) and [this](https://stackoverflow.com/questions/35069216/why-does-left-outer-join-not-use-the-index-correctly)). The solution was to re-structure the search query to use sub-selects that can benefit from the indexes.

The original query on the production database took around **3 minutes** https://explain.dalibo.com/plan/86628262f4f339a3
The original query on my local hashed community database took **4,5s** https://explain.dalibo.com/plan/2chf2dffc2cbha3d
The improved query with added gin indexes takes **65ms** https://explain.dalibo.com/plan/h3ha25g5g8bcfgbf

There is still some room for improvement, by adding [Full Text Search](https://www.postgresql.org/docs/current/textsearch-tables.html) to the work_package description field and rewriting the query to have a sub-select on the work package description condition. Considering the effort/reward of adding the FTS on the work package table, I would wait with it until it becomes a real problem. I would not go with a GIN index on the description field, because it can take a long time to re-index.

The problematic part of the work package query is the seq scan with filters applied:

<img width="633" alt="image" src="https://user-images.githubusercontent.com/83396/228465672-55018e0f-33bb-48bf-870c-fac949d38e6d.png">

The improvement with FTS would look like this https://explain.dalibo.com/plan/f5gcc2e477bbd4fe

<img width="947" alt="image" src="https://user-images.githubusercontent.com/83396/228469475-69f08350-8579-4ee8-b491-f258ecf0acce.png">

I would keep the work package subject and description for a later time, if the search query is still slow.
There is a potential drawback for the current gin index solution. If the journals table grows too big and gin indexing becomes slow on it, then some insert operations [could hang from time to time](https://gitlab.com/gitlab-com/gl-infra/production/-/issues/4725#note_596146675) if that happens then we can resort to FTS instead.

dombesz commented:
Follow up:

In the [original query](https://explain.dalibo.com/plan/86628262f4f339a3#) we also have slow clause on filtering the active projects. Added an index on the `projects.active` field to improve that clause as well.

<img width="420" alt="image" src="https://user-images.githubusercontent.com/83396/228501289-90b26e35-718e-42c8-9fb7-53c24a2101c2.png">

ulferts commented on db/migrate/20230329095313_add_index_on_projects.rb:
```diff
@@ -0,0 +1,5 @@
+class AddIndexOnProjects < ActiveRecord::Migration[7.0]
+  def change
+    add_index(:projects, :active)
```
I believe this index does not yield a performance gain. The only scenario in which I can see an index on a boolean field to improve the performance is when there is only a small portion of true/false values and that portion is searched for.

Admittedly, on some installations the number of inactive projects might outweigh the number of active projects (that are typically searched for). So it probably comes down to testing.

ulferts commented on db/migrate/20230328154645_add_gin_trgm_index_on_journals_and_custom_values.rb:
```diff
@@ -0,0 +1,25 @@
+    add_index(:journals, :notes, using: 'gin', opclass: :gin_trgm_ops)
+    add_index(:custom_values, :value, using: 'gin', opclass: :gin_trgm_ops)
+  rescue StandardError => e
+    raise unless e.message =~ /could not open extension/
```
I don't know if that is ever actually done but apparently it is possible to have error messages in languages other than English. In such a case, the check here would fail. 

It might be safer to restructure the change method so that the error handling is only done for the `enable_extension` part and the `add_index` part is only executed if the extension part is successful.

dombesz replied:
I extracted the method out, but just for safety I kept the condition and changed it to `e.message =~ /pg_trgm/`. The error message contains the name of the extension, and that should be the same, regardless of the OS's language.

ulferts commented on lib_static/plugins/acts_as_searchable/lib/acts_as_searchable.rb:
```diff
@@ -156,6 +141,58 @@ def searchable_projects_condition
+              custom_field_condition =
+                CustomValue.select('1').where(customized_type: name)
+                           .where("customized_id=#{table_name}.id AND value ILIKE ?")
+                           .where(custom_field_id: searchable_custom_field_ids)
```
This statement here (and I know that you just transformed it from the code before) is the [reason for the bug that list cfs cannot be searched by](https://community.openproject.org/projects/openproject/work_packages/34363).

If it makes sense to fix the bug within the PR or in a separate PR but right afterwards, feel free to @dombesz .

dombesz replied:
It'll be better to address the issue separately. I'll assign it to myself.

ulferts commented on lib_static/plugins/acts_as_searchable/lib/acts_as_searchable.rb:
```diff
@@ -83,41 +82,25 @@ module ClassMethods
-                                   " AND #{CustomValue.table_name}.custom_field_id IN (#{searchable_custom_field_ids.join(',')}))"
-                token_clauses << custom_field_sql
-              end
+              token_clauses += Array(searchable_custom_fields_conditions)
```
Every bit of improving the readability of that monster of a method is welcomed 👍

ulferts reviewed (changes requested):
The performance gains are reproducible and considerable. Well done @dombesz.

There are a couple of smaller issues to be fixed before the release but they are really minor.

dombesz commented:
@ulferts Thank you for the review, I'll addressed your comments. I would like to address the [bug](https://community.openproject.org/projects/openproject/work_packages/34363) you mentioned in a separate PR.

ulferts reviewed (approved):
👍
